### ISSUE TITLE
clang doesn't emit masked instructions from 256-bit AVX512VL mask intrinsics

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#37691]
|  |  |
| --- | --- |
| Bugzilla Link | [38343](https://llvm.org/bz38343) |
| Version | trunk |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @topperc,@chriselrod,@RKSimon,@rotateright |

## Extended Description 
Test case here:

https://godbolt.org/g/6apogL

Although the `mask_add256_avx512vl` function uses a mask intrinsic, it produces the same code (and I think the same IR) as the `mask_add256` function, which doesn't. (I wasn't able to get clang to emit masked ymm instructions at all).

**Comment by RKSimon** [comment]
> Codegen for non-constant masks:
> 
> https://gcc.godbolt.org/z/ixpSRz
> 
> Its appears to have a similar problem to [Bug #​33674]

Updated non-constant tests: https://gcc.godbolt.org/z/cnT5vod3r

**Comment by RKSimon** [comment]
Fixed in 10bb62319281c76827d6eda1ee70d4bac7e35935

Although one of the constant mask test cases still has the known problem that the select mask has widened from vXi32 to vXi64 preventing the fold:

```
__m512i mask_add512(__m512i a, __m512i b) {
    return _mm512_mask_add_epi32(a, 0xf0, a, b);
}

mask_add512(long long __vector(8), long long __vector(8)): # @mask_add512(long long __vector(8), long long __vector(8))
  vpaddd zmm1, zmm1, zmm0
  mov al, 12
  kmovd k1, eax
  vmovdqa64 zmm0 {k1}, zmm1
  ret

vs (icc)

mask_add512(__m512i, __m512i):
  mov eax, 240
  kmovw k1, eax
  vpaddd zmm0{k1}, zmm0, zmm1
  ret
```
